docs: add/update CLI plugin docs (#1539)
kubernetes-sigs/kubebuilder: CONTRIBUTING.md

```

1. Ensure you activate module support before continue (`$ export GO111MODULE=on`)
1. Run the command `make install` to create a bin with the source code

> 

Reviewer: spaces not part of the scope but 👍

Check the CI job after to do the Pull Request and then, click on in the `Details` of `netlify/kubebuilder/deploy-preview`

## Versioning

|   Name	|   Example	|  Description |
|---	|---	|---	|

> |  KubeBuilder version | `v2.2.0`, `v2.3.0`, `v2.3.1` | Tagged versions of the KubeBuilder project, representing changes to the source code in this repository. See the [releases][kb-releases] page for binary releases. |

Reviewer: /hold

Until we are able to update it related to the latest changes.

### Incrementing versions

Project versions should only be increased if a breaking change is introduced in the PROJECT file scheme itself. Changes to the Go scaffolding or the KubeBuilder CLI *do not* affect project version.

Similarly, the introduction of a new major version (`v(X+1).Y`) of the Go plugin might only lead to a new minor (`vX.(Y+1).Z`) release of KubeBuilder, since no breaking change is being made to the CLI itself. It'd only be a breaking change to KubeBuilder if we remove support for an older version of the plugin.

For more information on how the release and plugin versions work, see the [semver](https://semver.org/) documentation.

**NOTE:** the scheme for project version `"2"` was defined before the concept of plugins was introduced, so plugin `go.kubebuilder.io/v2.0` is implicitly used for those project types. Schema for project versions `"3-alpha"` and beyond define a `layout` key that informs the plugin system of which plugin to use.

## Introducing changes to plugins

> Changes made to plugins may require a plugin version increase depending on whether
> they break the current version or not. See the plugins design doc [versioning section][plugins-design-versioning]

Reviewer: ```suggestion
Changes that will break the expected behaviour of a project built with the previous _minor_ Plugin versions is a breaking change to that plugin.  See the plugins design doc [versioning section][plugins-design-versioning]

```

for more details.

Plugin changes only warrant a new KubeBuilder release if and only if a plugin
version is no longer supported by the CLI.

### Example

KubeBuilder release version `v2.3.1` scaffolds projects with plugin `go.kubebuilder.io/v2.1`
by default. Previously, plugin `go.kubebuilder.io/v2.0` was used.

> You make a PR that adds a new marker to the file `main.go` scaffolded by `init`

Reviewer: ```suggestion
Let's imagine that our PR adds a new marker to the file `main.go` scaffolded by `init`
```

that `create api` will use to update that file. The changes introduced in your PR

> will cause errors when used with projects built with plugins `go.kubebuilder.io/v2.0`

Reviewer: ```suggestion
will cause errors when used with projects built with plugins `go.kubebuilder.io/v3.0`
```

and `go.kubebuilder.io/v2.1` without users manually updating their projects.

> Thus, your changes introduce a breaking change to plugin `go.kubebuilder.io`,

Reviewer: ```suggestion
Thus, our changes introduce a breaking change to plugin `go.kubebuilder.io`,
```

> You make a PR that adds a new marker to the file `main.go` scaffolded by `init`
> that `create api` will use to update that file. The changes introduced in your PR
> will cause errors when used with projects built with plugins `go.kubebuilder.io/v2.0`
> and `go.kubebuilder.io/v2.1` without users manually updating their projects.
> Thus, your changes introduce a breaking change to plugin `go.kubebuilder.io`,
> and require a _major_ version increase to `v3.0` in the PR.

Reviewer: ```suggestion
Let's imagine that our PR adds a new marker to the file `main.go` scaffolded by `init`
that `create api` will use to update that file. The changes introduced in our PR
will cause errors when used with projects which were built with plugins `go.kubebuilder.io/v3.0`
and `go.kubebuilder.io/v3.1` without users manually updating their projects.
Thus, our changes introduce a breaking change to plugin `go.kubebuilder.io` and then,
the Plugin requires a _major_ version increase to `v4.0` in the PR.
```

> You must also add a migration guide to the [migrations](https://book.kubebuilder.io/migrations.html)
> section of the KubeBuilder book in your PR. It should detail the steps required
> for users to upgrade their projects from `go.kubebuilder.io/v2.Y` to the new
> plugin version `v3.0`.

Reviewer: ```suggestion
We must also add a migration guide to the [migrations](https://book.kubebuilder.io/migrations.html)
section of the KubeBuilder book in your PR. It should detail the steps required
for users to upgrade their projects from `go.kubebuilder.io/v3.Y` to the new
plugin version `v4.0`.
```

> You've only added a new plugin version without removing old versions, so this

Reviewer: ```suggestion
Note that, we've only added a new plugin version without removing old versions, so this
```

> considered a new feature to KubeBuilder and not a breaking change. Therefore
> a new _minor_ KubeBuilder version `v2.4.0` should be release.
> 

Reviewer: The explanation here lead us to believe that after each change we need to do a release, which shows not true at all.
```suggestion
considered a new feature to KubeBuilder and not a breaking change. Therefore
if we decide to release a new version of KubeBuilder then, a new _minor_ KubeBuilder version with the tag `v5.4.0` should be done.
```

Check the CI job after to do the Pull Request and then, click on in the `Details` of `netlify/kubebuilder/deploy-preview`

## Versioning

|   Name	|   Example	|  Description |
|---	|---	|---	|

> |  KubeBuilder version | `v2.2.0`, `v2.3.0`, `v2.3.1` | Tagged versions of the KubeBuilder project, representing changes to the source code in this repository. See the [releases] page for binary releases. |

Reviewer: Missing the link

|  Project version |  `"1"`, `"2"`,`"3-alpha"` | Project version defines the scheme of a `PROJECT` configuration file. This version is defined in a `PROJECT` file's `version`. |

> |  Plugin version | `v2.0`, `v2.1` | Represents the version of an individual plugin, as well as the corresponding scaffolding that it generates. This version is defined in a plugin key, ex. `go.kubebuilder.io/v2.0`. See the [design doc][plugins-design-versioning] for more details. |

Reviewer: Currently, we have implemented v2.0.0. So, why change for v2.0?
Otherwise, we are unable to bum v2.0.1, for example, which represents a change that is NOT a breaking change and is just a bugfix. Was not it that you indeed suggested in the latest KB meeting?

Author: See [this conversation](https://github.com/kubernetes-sigs/kubebuilder/issues/1519#issuecomment-632334210), which discusses dropping minor version from plugin versions.

Reviewer: OK. I am ok with we move in this way.

|  Plugin Versions | `go.kubebuilder.io/v2.0.0` | Represents the version of an individual plugin, as well as the corresponding scaffolding that it generates. |
|  KubeBuilder version | `v2.2.0`, `v2.3.0`, `v2.3.1` | Tagged versions of the KubeBuilder project, representing changes to the source code in this repository. See the [releases] page for binary releases. |
|  Project version |  `"1"`, `"2"`,`"3-alpha"` | Project version defines the scheme of a `PROJECT` configuration file. This version is defined in a `PROJECT` file's `version`. |
|  Plugin version | `v2.0`, `v2.1` | Represents the version of an individual plugin, as well as the corresponding scaffolding that it generates. This version is defined in a plugin key, ex. `go.kubebuilder.io/v2.0`. See the [design doc][plugins-design-versioning] for more details. |

> Note that PROJECT version should only be bumped if a breaking change is introduced in the PROJECT file format itself.  Changes to the Go scaffolding or the KubeBuilder CLI *do not* affect the PROJECT version.

Reviewer: Why remove this information?

For more information on how the release and plugin versions work, see the [semver](https://semver.org/) documentation.

> **NOTE:** schema for project versions `"1"` and `"2"` were defined before the concept of plugins was introduced, so plugin `go.kubebuilder.io/v2.0` is implicitly used for those project types. Schema for project versions `"3-alpha"` and beyond define a `layout` key that informs the plugin system of which plugin to use.

Reviewer: `plugin `go.kubebuilder.io/v2.0` is implicitly used for those project types` -> It is not true since it ONLY represents the layout of the files scaffolded by default in V2 projects. V1 projects have a different layout and following the same analogy would be `go.kubebuilder.io/v1.0` besides its code source no longer exist in the project.

Author: That is incorrect, they both use(d) `go.kubebuilder.io/v2.0.0`.

Author: Ah yes there was a `v1.0.0` plugin. We should just remove discussion of project version 1 to avoid confusion, since we don't talk about it anywhere else.

Reviewer: Agreed .. Remove V1 from here as well.

> **NOTE:** schema for project versions `"1"` and `"2"` were defined before the concept of plugins was introduced, so plugin `go.kubebuilder.io/v2.0` is implicitly used for those project types. Schema for project versions `"3-alpha"` and beyond define a `layout` key that informs the plugin system of which plugin to use.

Reviewer: IMO we should keep X.X.X always  ( have the possibility of us X.X or X.X.X increase the complexity and can lead from wrong understanding such as that we should not bump the minor for bugfixes )

Author: https://github.com/kubernetes-sigs/kubebuilder/pull/1539#discussion_r433429664

## Introducing changes to plugins

Changes made to plugins may require a plugin version increase depending on whether

> they break the current version or not. See the plugins design doc [versioning section][plugins-design-versioning]

Reviewer: The above line brings unnecessary complexity  see: 

You replace a concise definition rule ` any change that will break the expected behaviour of a project built with the previous `MINOR` Plugin versions is a breaking change to that plugin. `

BY : 

`Changes made to plugins may require a plugin version increase depending on whether they break the current version or not` 

Which makes this guidance very subjective. Note that the doc/link do not define and/or clarifies what means a breaking change for the plugins too. I am ok with add the link but I prefer the previous definition which makes the rule very clear and objective.

Author: I didn't replace this sentence, I moved it to the plugins design doc.

Reviewer: The doc defines the proposal solution, but I am not sure if should be used as a reference to define what is not breaking cange and how to contribute with.  I would prefer to keep a simple and no verbose explanation here as before.

for more details.

Breaking changes require either a new Kubebuilder _minor_ or _major_ release.
A new _major_ release is only necessary for changes that fundamentally alter

> all `kubebuilder` projects.

Reviewer: The above is very confusing for me. See: 

`Breaking changes require either a new Kubebuilder _minor_ or _major_ release`

If we are following the semantic version convention it shows make no sense. How a breaking change can require either? What does it means? 

`A new _major_ release is only necessary for changes that fundamentally alter
all `kubebuilder` projects.`

What do you mean with  `kubebuilder` projects.` here? What is a change that fundamentally alter all `kubebuilder` projects.?

Author: >What means a change that fundamentally alter all kubebuilder projects.?

I agree this is a little nebulous and could use some refinement. Thoughts @DirectXMan12 @joelanford?

Reviewer: The doc defines the proposal solution, but I am not sure if should be used as a reference to define what is not breeding cange and how to contribute with.  I would prefer to keep a simple and no verbose explanation here as before.

Author: We have a concrete definition of breaking plugin changes now, and no discussion of releasing kubebuilder in relation to plugins.

> Non-breaking changes do not require a release.

Reviewer: ```suggestion
Non-breaking changes do not require a release.
```
It makes no sense for me since the releases are not ONLY done when has breaking changes in the plugins Actually, has or not breaking changes in the plugins has no relation with the releases.

Author: I think this should be "if you remove a plugin major version, you must release a kubebuilder major version".

### Example

KubeBuilder release version `v2.3.1` scaffolds projects with plugin `go.kubebuilder.io/v2.1`

> by default. Previously, plugin `go.kubebuilder.io/v2.0` was used.

Reviewer: I am afraid that use the `v2.3.1` which is a valid release brings confusing since because of this it the example was using a version that was not released yet to illustrate how it works.

Author: Personally I don't think the versions we use in an example matter.

Reviewer: The text `KubeBuilder release version `v2.3.1` scaffolds projects with plugin `go.kubebuilder.io/v2.1` shows a fact because the release and version exists and lead the impression that the v2.X.X would just release plugins v2.X when both are not true. So, I'd suggest:

Author: Removed kubebuilder version from this section.

Participation in the Kubernetes community is governed by the [Kubernetes Code of Conduct](code-of-conduct.md).

> [kb-releases]:ttps://github.com/kubernetes-sigs/kubebuilder/releases

Reviewer: ```suggestion
[kb-releases]:https://github.com/kubernetes-sigs/kubebuilder/releases
```

Participation in the Kubernetes community is governed by the [Kubernetes Code of Conduct](code-of-conduct.md).

[releases]:ttps://github.com/kubernetes-sigs/kubebuilder/releases

> [plugins-design-versioning]:design/extensible-cli-and-scaffolding-plugins-phase-1.md

Reviewer: The links are broken.

Author: Should be fixed.
